src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -881,21 +1374,136 @@ void W3DDisplay::Preload_Texture_Assets(Utf8String texture)
+
+#define DibSize(lpbi) ((lpbi)->biSize + (lpbi)->biSizeImage + (int)(lpbi)->biClrUsed * sizeof(RGBQUAD))
+
+#define DibPaletteSize(lpbi) (DibNumColors(lpbi) * sizeof(RGBQUAD))
+void Create_Bmp_File(const char *filename, const char *data, int width, int height)
+{
+#ifdef BUILD_WITH_D3D8
+    BITMAPINFOHEADER *header = (BITMAPINFOHEADER *)LocalAlloc(LMEM_ZEROINIT, sizeof(BITMAPINFOHEADER));
+    header->biSize = sizeof(BITMAPINFOHEADER);
+    header->biWidth = width;
+    header->biHeight = height;
+    header->biPlanes = 1;
+    header->biBitCount = 24;
+    header->biCompression = 0;
+    header->biSizeImage = 24 * header->biHeight * ((header->biWidth + 7) / 8);
+    header->biClrImportant = 0;
+    HANDLE h = CreateFile(filename, GENERIC_READ | GENERIC_WRITE, 0, nullptr, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr);
+
+    if (h != INVALID_HANDLE_VALUE) {
+        BITMAPFILEHEADER fheader;
+        fheader.bfType = BFT_BITMAP;
+        fheader.bfSize = DibSize(header) + sizeof(BITMAPFILEHEADER);
+        fheader.bfReserved1 = 0;
+        fheader.bfReserved2 = 0;
+        fheader.bfOffBits = header->biSize + DibPaletteSize(header) + sizeof(BITMAPFILEHEADER);
+        DWORD NumberOfBytesWritten;
+
+        if (WriteFile(h, &fheader, sizeof(BITMAPFILEHEADER), &NumberOfBytesWritten, 0)) {
+            if (WriteFile(h, header, 4 * header->biClrUsed + sizeof(BITMAPINFOHEADER), &NumberOfBytesWritten, 0)) {
+                if (WriteFile(h, data, header->biSizeImage, &NumberOfBytesWritten, 0)) {
+                    CloseHandle(h);
+                }
+            }
+        }
+    }
+
+    // BUGFIX prevent memory leak;
+    LocalFree(header);
+#endif
+}
+
 // 0x00740D40
 void W3DDisplay::Take_ScreenShot()
 {
-#ifdef GAME_DLL
-    Call_Method<void, W3DDisplay>(0x00740D40, this);
+#ifdef BUILD_WITH_D3D8
+    bool found = false;
+    static int currentShot = 1;
+    char fname[1024];
+    char buf[256];
+
+    while (!found) {
+        sprintf(buf, "%s%.3d.bmp", "sshot", currentShot);
+        currentShot++;
+        strcpy(fname, g_theWriteableGlobalData->m_userDataDirectory);
+        strcat(fname, buf);
+
+        if (_access(fname, 0) != -1) {
+            found = true;
+        }
+    }
+
+    IDirect3DSurface8 *surface = DX8Wrapper::Get_DX8_Front_Buffer();
+    D3DSURFACE_DESC desc;
+    surface->GetDesc(&desc);
+    RECT rect;
+    POINT point;
+    GetClientRect(g_applicationHWnd, &rect);
+
+    point.x = rect.left;
+    point.y = rect.top;
+    ClientToScreen(g_applicationHWnd, &point);
+    rect.left = point.x;
+    rect.top = point.y;
+
+    point.x = rect.right;
+    point.y = rect.bottom;
+    ClientToScreen(g_applicationHWnd, &point);
+    rect.right = point.x;
+    rect.bottom = point.y;
+
+    D3DLOCKED_RECT locked_rect;
+    DX8Wrapper::Handle_DX8_ErrorCode(surface->LockRect(&locked_rect, &rect, D3DLOCK_READONLY));
+
+    unsigned int x = rect.right - rect.left;
```

Comment:
width and height would be a better name as x/y suggests they iterate. (then you can use x and y for the next loops)